crates/ibc/src/core/ics03_connection/handler/conn_open_confirm.rs

```diff
@@ -1,15 +1,174 @@
 //! Protocol logic specific to processing ICS3 messages of type `MsgConnectionOpenConfirm`.
 
+use crate::core::context::ContextError;
 use crate::core::ics03_connection::connection::{ConnectionEnd, Counterparty, State};
 use crate::core::ics03_connection::context::ConnectionReader;
 use crate::core::ics03_connection::error::ConnectionError;
 use crate::core::ics03_connection::events::OpenConfirm;
 use crate::core::ics03_connection::handler::{ConnectionIdState, ConnectionResult};
 use crate::core::ics03_connection::msgs::conn_open_confirm::MsgConnectionOpenConfirm;
+use crate::core::ics24_host::identifier::{ClientId, ConnectionId};
+use crate::core::ics24_host::path::ConnectionsPath;
+use crate::core::{ExecutionContext, ValidationContext};
 use crate::events::IbcEvent;
 use crate::handler::{HandlerOutput, HandlerResult};
 use crate::prelude::*;
 
+pub(crate) fn validate<Ctx>(ctx_b: &Ctx, msg: &MsgConnectionOpenConfirm) -> Result<(), ContextError>
+where
+    Ctx: ValidationContext,
+{
+    let vars = LocalVars::new(ctx_b, msg)?;
+    validate_impl(ctx_b, msg, &vars)
+}
+
+fn validate_impl<Ctx>(
+    ctx_b: &Ctx,
+    msg: &MsgConnectionOpenConfirm,
+    vars: &LocalVars,
+) -> Result<(), ContextError>
+where
+    Ctx: ValidationContext,
+{
+    let conn_end_on_b = vars.conn_end_on_b();
+    if !conn_end_on_b.state_matches(&State::TryOpen) {
+        return Err(ContextError::ConnectionError(
+            ConnectionError::ConnectionMismatch {
+                connection_id: msg.conn_id_on_b.clone(),
+            },
+        ));
+    }
+
+    let client_id_on_a = vars.client_id_on_a();
+    let client_id_on_b = vars.client_id_on_b();
+    let conn_id_on_a = vars.conn_id_on_a()?;
+
+    // Verify proofs
+    {
+        let client_state_of_a_on_b =
+            ctx_b
+                .client_state(client_id_on_b)
+                .map_err(|_| ConnectionError::Other {
+                    description: "failed to fetch client state".to_string(),
+                })?;
+        let consensus_state_of_a_on_b = ctx_b
+            .consensus_state(client_id_on_b, msg.proof_height_on_a)
+            .map_err(|_| ConnectionError::Other {
+                description: "failed to fetch client consensus state".to_string(),
+            })?;
```

Comment:
Same here and elsewhere, where we call context methods that return `ContextError` ->
```suggestion
            .consensus_state(client_id_on_b, msg.proof_height_on_a)?;
```

Reply:
Same reasoning here